gen_initramfs.sh

```diff
@@ -1706,15 +1706,44 @@ append_firmware() {
 
 	mkdir -p "${TDIR}"/lib/firmware || gen_die "Failed to create '${TDIR}/lib/firmware'!"
 
-	if [ ${#FIRMWARE_FILES[@]} -gt 0 ]
+	if isTrue "${ALLFIRMWARE}"
+	then
+		cp -a "${FIRMWARE_DIR}"/* "${TDIR}"/lib/firmware/ 2>/dev/null \
+			|| gen_die "Failed to copy firmware files to '${TDIR}/lib/firmware'!"
+	elif [ ${#FIRMWARE_FILES[@]} -gt 0 ]
 	then
 		pushd "${FIRMWARE_DIR}" &>/dev/null || gen_die "Failed to chdir to '${FIRMWARE_DIR}'!"
 		cp -rL --parents --target-directory="${TDIR}/lib/firmware" "${FIRMWARE_FILES[@]}" 2>/dev/null \
 			|| gen_die "Failed to copy firmware files (${FIRMWARE_FILES}) to '${TDIR}/lib/firmware'!"
 		popd &>/dev/null || gen_die "Failed to chdir!"
 	else
-		cp -a "${FIRMWARE_DIR}"/* "${TDIR}"/lib/firmware/ 2>/dev/null \
-			|| gen_die "Failed to copy firmware files to '${TDIR}/lib/firmware'!"
+		local myfw=
+		local -a fwlist=()
+		while IFS= read -r -u 3 myfw
+		do
+			if [ -z "${myfw}" ]
+			then
+				continue
+			fi
+
+			if [ ! -f "${FIRMWARE_DIR}/${myfw}" ]
+			then
+				print_warning 3 "$(get_indent 3) - ${myfw} is missing; Ignoring ..."
+				continue
+			fi
+
+			fwlist+=( "${myfw}" )
+		done 3< <( (
+			modinfo -b "${KERNEL_MODULES_PREFIX%/}" -k "${KV}" -F firmware ${GK_INITRAMFS_KMODULES} 2>/dev/null
+		) | sort | uniq )
+
+		if [ ${#fwlist[@]} -gt 0 ]
+		then
+			pushd "${FIRMWARE_DIR}" &>/dev/null || gen_die "Failed to chdir to '${FIRMWARE_DIR}'!"
```

Comment:
this block here is almost identical to `cp -rL --parents --target-directory=...` above, seems like a good candidate for refactoring (so then it gets fwlist[@] or FIRMWARE_FILES[@] as the inputs.